proposed/2023/release-staging-and-deprecation.md

```diff
@@ -0,0 +1,88 @@
+For example, all packages published as part of .NET 8 Preview 5 release will be staged together and then also can be deprecated together by a single call to NuGet APIs.
+
+## Motivation 
+
+<!-- Why are we doing this? What pain points does this solve? What is the expected outcome? -->
+Many products (Apps / Tools / Frameworks / Utilities / etc. ...) consist of multiple interdependent packages which should be published together as a single set. Currently, NuGet provides no functionality for definding / grouping packages into such sets. This results in package owners having to keep track of these package sets on their own. There are three specific sets of scenarios which this feature will simplify dramatically:
+1. When preparing a release, product owners would like to have a mechanism to pre-stage their product packages, so that the release will consist only of a single action of marking the set of packages as `published`, which would result all the packages in the set to be marked as published at the same time.
+2. Sometimes developers have to take a dependency on several related packages in their projects, which are part of the same "release". Without having a concept of the `release` developers have to guess which versions of those packages are compatible. As a result, most of hte time developers would simply pick the latest version of all these packages, which somtimes can lead to incompatibility issues down the road.
+3. Keeping a package healthy on NuGet also means keeping the list of packages aligned with it support policy. Hence, package owners try to deprecate packages which are out of support. Without ability to group packages, package owners end up keeping track of all the packages which belong to the same release and run the deprecation over a discrete set of packages when necessary. Being able to group packages together would enable deprecating a whole release at once, knowing that all the packages which belong to that release will get deprecated.
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+There is a new field in the package metadata, which indicates the unique release of the product the package belongs to. This field is called `release` and the value is scoped to the publisher of the package.
+
+When a package is being built, that field in the metadata is set to the same value for all the related packages, that are built together.
+
+Now, with this information in place, new set of NuGet APIs / extensions allow the user to run the following actions:
+- **Search / Query for packages based on a release** for a given publisher
+- **Deprecate a relase** for current publisher
+- **Stage** a package for later publishing
+- **Publish an earlier staged release**
```

Comment:
Say a package ID that has never been published before is pre-staged. At what point should the the package ID be claimed, preventing any other account from pubishing a package with the same ID?  Or should each server be free to implement whatever they feel is best?

Reply:
I assume what you refer here is not package ID, but the new metadata I introduced - "release".
I'm having hard time understand the concept of different `servers`. Maybe I lack some critical knowledge here. I'll schedule a quick meeting to go over that with you and maybe then I'll be able to better answer this question.

For now, I assume something like this can work:
1. Staging request on its own can be that signal. So that when the publisher is about to upload a set of packages for staging, they first call some `stage release {ReleaseID}` API, passing in the release they'd like to use.
2. They receive back some "key" in response to the call above
3. For each package upload they provide the key as a way to associate the package with the release.

This addresses the issue of somebody else using the same release to upload another package, because they don't have the unique key, which will be considered a temporary secret.

Reply:
At what point the package ID is an important question, but I think it's easy. It should be reserved when the package enters the staging area, not later when the staging area is published as a batch. Any concerns of locking up too many package IDs in the stage should be mitigated with throttling or rate-limiting activities done automatically today or handled on a case-by-case basis (many abuse/fraud cases are special and need human eyes to mitigate). 

Having the package ID reserved only later would reduce the usefulness of the staging area. There should be as few reasons as possible for the staging area to fail publishing. Package metadata validation, signing validation, ownership checks, etc etc all should happen as early as possible so the publish of a staging area (i.e. release) is a smooth operation.

Reply:
Agreed with @joelverhagen here, and making a note in the proposal about this.